cpp/cert/test/rules/ERR55-CPP/test_no_except.cpp

```diff
@@ -1,9 +1,34 @@
-#include <stdexcept>
+class ExceptionA {};
 
-void test_noexcept_false() { // COMPLIANT
-  throw "test";
+void test_throw() noexcept(true) {
+  throw ExceptionA(); // NON_COMPLIANT - function marked as noexcept(true)
 }
 
-void test_noexcept_true() noexcept(true) { // NON_COMPLIANT
-  throw "test";
+void throwA() { throw ExceptionA(); }
+void indirectThrowA() { throwA(); }
+void noexceptIndirectThrowA() noexcept { throwA(); } // NON_COMPLIANT
+
+void test_indirect_throw() noexcept(true) {
+  throwA(); // NON_COMPLIANT - function marked as noexcept(true)
+}
+
+void test_indirect_throw_2() noexcept {
+  throwA(); // NON_COMPLIANT - function marked as noexcept(true)
+}
+
+void test_indirect_throw_3() noexcept(false) {
+  throwA(); // COMPLIANT - function marked as noexcept(false)
+}
+
+void test_indirect_throw_4() {
+  throwA(); // COMPLIANT - function marked as noexcept(false)
+}
+
+void test_indirect_throw_5() noexcept {
+  noexceptIndirectThrowA(); // COMPLIANT - noexceptIndirectThrowA would call
+                            // std::terminate() if ExceptionA is thrown
+}
+
+void test_indirect_throw_6() noexcept {
+  indirectThrowA(); // NON_COMPLIANT
 }
```

Comment:
```suggestion
}

void test_indirect_throw_6() noexcept {
    dynamicIndirectThrowA(); // NON_COMPLIANT[FALSE_NEGATIVE] - dynamicIndirectThrowA indirectly throws std::bad_exception
}
```

Reply:
Ah, this is actually not a false negative. `dynamicIndirectThrowA` will not throw a `std::bad_exception` unless:

  * [`std::set_unexpected`](http://en.cppreference.com/w/cpp/error/set_unexpected) has been called with a custom unexpected error handler (note: this handler functionality was deprecated in C++11 and removed in C++17).
  * The custom handler re-throws an exception, and the exception it rethrows is not compatible with the dynamic exception specification.

This is shown in this example here: https://en.cppreference.com/w/cpp/error/bad_exception. Given this relates to deprecated functionality in our supported C++ standards (C++14 onwards), I don't think we need to implement this edge case or add a test for it. Let me know if you disagree.

Note: the [ERR55-CPP](https://wiki.sei.cmu.edu/confluence/display/cplusplus/ERR55-CPP.+Honor+exception+specifications) CERT rule has an incorrect statement about this behaviour:

> If a function declared with a dynamic-exception-specification throws an exception of a type that would not match the exception-specification, the function std::unexpected() is called **The behavior of this function can be overridden but, by default, causes an exception of std::bad_exception to be thrown**

That is not true - the default behaviour is for it to call `std::terminate`. You have to set the `unexpected` handler to trigger any of this behaviour.

Reply:
Agreed, it is only a FP in the case of a custom unexpected handler that rethrows an unspecified exception type. Good catch on the mistake in CERT!